compiler/rustc_trait_selection/src/solve/normalize.rs

```diff
@@ -41,6 +42,19 @@ pub(crate) fn deeply_normalize_with_skipped_universes<'tcx, T: TypeFoldable<TyCt
     value.try_fold_with(&mut folder)
 }
 
+// Deeply normalize a value and return it
+pub(crate) fn deeply_normalize_for_diagnostics<'tcx, T: TypeFoldable<TyCtxt<'tcx>>>(
+    infcx: &InferCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    t: T,
+) -> T {
+    infcx
+        .commit_if_ok(|_| {
+            deeply_normalize(infcx.at(&ObligationCause::dummy(), param_env), t.clone())
```

Comment:
this would keep `<() as CanNormalize>::Assoc: Trait<<() as Ambig>::Assoc>` fully unnormalized. Maybe easiest to write a `TypeFolder` here where we call `deeply_normalize` and if that returns an error, `super_fold_with`?